### PR TITLE
Day 3/sale

### PR DESCRIPTION
**Pull request by salac-kolosek**

### CODE REVIEW

**Comment by kristina-kolosek** [comment]
Before I check the code, here's functional issues I found:

* Api key page should redirect to root, if api key is already set
* I can't see full root page, I see like the middle of it and can't scroll up or down, fix that
* When a page containing a build is refreshed, I get an error. It should be able to refetch the data from API for single page and display the build

File: src/containers/CircleCI/ApiProvider.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+class ApiProvider extends Component {
+
+  state = { 
+    input: '',
+    redirect: false, 
+  }
+
+  componentDidMount() {
+    const apiKey = sessionStorage.getItem('CIRCLE_CI_API_KEY');
+    this.setState({redirect: apiKey});
```
Comment (kristina-kolosek) [inline src/containers/CircleCI/ApiProvider.js]: better use this instead:
  this.setState({redirect: !!apiKey});

This way, you'll always for sure have true/false value in state

File: src/containers/CircleCI/ApiProvider.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  handleSave = () => {
+    sessionStorage.setItem('CIRCLE_CI_API_KEY', this.state.input)
```
Comment (kristina-kolosek) [inline src/containers/CircleCI/ApiProvider.js]: You're not checking whether api key is valid when saving it

File: src/containers/CircleCI/List.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+class List extends Component {
+
+  state = { 
+    redirect: false, 
+  }
+
+  componentDidMount() {
+    const apiKey = sessionStorage.getItem('CIRCLE_CI_API_KEY');
+    this.props.getBuildList(apiKey);
```
Comment (kristina-kolosek) [inline src/containers/CircleCI/List.js]: why even call API if you know there's no api key?

File: src/containers/CircleCI/List.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import './css/List.css'
+
+class List extends Component {
+
+  state = { 
+    redirect: false, 
+  }
+
+  componentDidMount() {
+    const apiKey = sessionStorage.getItem('CIRCLE_CI_API_KEY');
```
Comment (kristina-kolosek) [inline src/containers/CircleCI/List.js:22]: It's a good idea to connect the api key in session storage to redux store, so you don't always access session storage

File: src/containers/CircleCI/List.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            return this.renderSingleBuild(build)
+          })
+        }
+      </div>
+    )
+  }
+
+  renderSingleBuild(build) {
+    return (
+      <div key={uuidv1()} className="single-build">
```
Comment (kristina-kolosek) [inline src/containers/CircleCI/List.js]: This should be a separate component

Comment (kristina-kolosek) [reply src/containers/CircleCI/List.js]: Also, don't use uuidv1 for key as it always returns other key, meaning that the whole list will always be rerendered when render is called, even if there's no need for that:
key should be unique, yet persistent per item

File: src/reducers/CircleCIReducer/actions.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      })
+  }
+}
+
+export const getBuild = (buildNum, apiKey) => {
+  return (dispatch) => {
+    dispatch({
+      type: action_types.GET_BUILD_START
+    });
+    return axios.get('https://circleci.com/api/v1.1/recent-builds?circle-token=' + apiKey)
```
Comment (kristina-kolosek) [inline src/reducers/CircleCIReducer/actions.js]: idea was to get a build directly, not to filter a list - there's api call for this + it's more optimized

**Review by kristina-kolosek — CHANGES_REQUESTED** [review]
Besides requested changes, also proceed with bonus task

File: src/containers/CircleCI/ApiProvider.js (hunk 1 above)
Comment (salac-kolosek) [reply src/containers/CircleCI/ApiProvider.js]: Ok, tnx for info :)

**Review by kristina-kolosek — CHANGES_REQUESTED** [review]
You didn't do the bonus task, but code looks fine

**Comment by kristina-kolosek** [comment]
Did you ever do bonus task in the end? :)

**Comment by salac-kolosek** [comment]
I didn't. I can do it tomorrow if I get some time for it :)

**Comment by kristina-kolosek** [comment]
It's okay, I was kidding :) closing this, please make a PR for day 4

**Pull request closed without merging**